lib/scheduler/thread_pool.rb

```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+module Scheduler
+  # ThreadPool manages a pool of worker threads that process tasks from a queue.
+  # It maintains a minimum number of threads and can scale up to a maximum number
+  # when there's more work to be done.
+  class ThreadPool
+    class ShutdownError < StandardError
+    end
+
+    def initialize(min_threads:, max_threads:, idle_time:)
+      raise ArgumentError, "min_threads must be positive" if min_threads <= 0
```

Comment:
I'm wondering why we don't allow `min_threads: 0` since we may not want to keep a thread running when there is no work for an extended period of time.